traits/src/ff/mod.rs

```diff
@@ -0,0 +1,194 @@
+    + for<'a> core::ops::Add<&'a Self, Output = Self>
+    + for<'a> core::ops::Sub<&'a Self, Output = Self>
+    + for<'a> core::ops::Mul<&'a Self, Output = Self>
+    + for<'a> core::iter::Sum<&'a Self>
+    + for<'a> core::iter::Product<&'a Self>
+    + core::ops::AddAssign
+    + core::ops::SubAssign
+    + core::ops::MulAssign
+    + for<'a> core::ops::AddAssign<&'a Self>
+    + for<'a> core::ops::SubAssign<&'a Self>
+    + for<'a> core::ops::MulAssign<&'a Self>
+    + From<u64>
+{
+    /// The zero element of the field, the additive identity.
+    const ZERO: Self;
+
+    /// The one element of the field, the multiplicative identity.
+    const ONE: Self;
+
+    /// Returns an element chosen uniformly at random using a user-provided RNG.
+    fn random(rng: impl RngCore) -> Self;
+
+    /// Returns true iff this element is zero.
+    fn is_zero(&self) -> subtle::Choice {
+        self.ct_eq(&Self::ZERO)
+    }
+
+    /// Interpret a string of decimal digits as a (congruent) field element.
+    fn from_decimal_str(s: &str) -> Option<Self> {
+        crate::ff::utils::from_str(s, 10)
+    }
+
+    /// Interpret a string of hex digits as a (congruent) field element.
+    fn from_hex_str(s: &str) -> Option<Self> {
+        crate::ff::utils::from_str(s, 16)
+    }
+
+    /// Squares this element.
+    #[must_use]
+    fn square(&self) -> Self {
+        *self * self
+    }
+
+    /// Doubles this element.
+    #[must_use]
+    fn double(&self) -> Self {
+        *self + self
+    }
+
+    /// Returns the square root of the field element, if it is
+    /// quadratic residue.
+    fn sqrt(&self) -> subtle::CtOption<Self>;
+
+    /// Computes the multiplicative inverse of this element,
+    /// failing if the element is zero.
+    fn invert(&self) -> subtle::CtOption<Self>;
+
+    /// Exponentiates `self` by `exp`, where `exp` is a little-endian order integer
+    /// exponent.
+    ///
+    /// # Guarantees
+    ///
+    /// This operation is constant time with respect to `self`, for all exponents with the
+    /// same number of digits (`exp.as_ref().len()`). It is variable time with respect to
+    /// the number of digits in the exponent.
+    fn pow<S: AsRef<[u64]>>(&self, exp: S) -> Self {
+        crate::ff::arithmetic::pow(self, exp)
+    }
+
+    /// Exponentiates `self` by `exp`, where `exp` is a little-endian order integer
+    /// exponent.
+    ///
+    /// # Guarantees
+    ///
+    /// **This operation is variable time with respect to `self`, for all exponent.** If
+    /// the exponent is fixed, this operation is effectively constant time. However, for
+    /// stronger constant-time guarantees, [`Field::pow`] should be used.
+    fn pow_vartime<S: AsRef<[u64]>>(&self, exp: S) -> Self {
+        crate::ff::arithmetic::pow_vartime(self, exp)
+    }
+
+    /// Returns a field element that is congruent to the provided little endian unsigned
+    /// byte representation of an integer.
+    fn from_uniform_bytes(bytes: &[u8]) -> Self;
```

Comment:
From a discussion with David I learned that a tower of fields is also a field.  When our field is a tower probably it doesn't make sense to think about an integer representation (it would be several integers).  The same applies to the `From<u64>` that this trait has.

Maybe the `From<u64>` and `from_uniform_bytes` should be moved to `PrimeField`?

Reply:
`PrimeField` had to be implemented for `Fq2` because `Repr` is in `PrimeField` in `zkcrypto/ff`. Maybe we no longer have that need. But still note that towers are also `PrimeField` and

We actually need `from_uniform_bytes` in Fq2 for hash to g2 curve. But I get your point if we can manage to construct Fq2 with two Fqs in hash to curve suite so it's fine to move it to `PrimeField`. Another point is that I'm not sure if it counts but with `zkcrypto/ff` we have to import `PrimeField` in halo2 everywhere because `from_uniform_bytes` and `Repr` are in `PrimeField` so now they all could be `Field` traits.

Reply:
>  Another point is that I'm not sure if it counts but with `zkcrypto/ff` we have to import `PrimeField` in halo2 everywhere because `from_uniform_bytes` and `Repr` are in `PrimeField` so with this change many stuff could be turned to `Field` trait.

Hm, perhaps we should really use `PrimeField` everywhere in halo2 instead of `Field`?  As an example, this is the "`Field`" we're using in zkevm-circuits https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/e7171f6601c236da05253ef36113149817f8b6d6/eth-types/src/lib.rs#L102-L106

Reply:
@ed255 should we also consider other types of fields especially at frontend side of halo2? and ideally that Field trait should cover a common ground?